From 948b544a088e3c6e6077ba837035ada38d072b5a Mon Sep 17 00:00:00 2001
From: Lawrence Rust <lvr@softsystem.co.uk>
Date: Sun, 20 Jul 2014 11:33:54 +0100
Subject: [PATCH 257/333] FE: Add a UI to display and adjust a recording's
 star rating

This adds a ratingstate control to the watchrecordings window below
the recording's descriptive text.  The user can adjust the star rating
by using the 7 and 9 keys, as used im mythmusic.

Signed-off-by: Lawrence Rust <lvr@softsystem.co.uk>
---
 mythtv/libs/libmyth/programinfo.cpp             |   21 +++++++++++++
 mythtv/libs/libmyth/programinfo.h               |    1 +
 mythtv/libs/libmythtv/tv_play.cpp               |    4 +++
 mythtv/programs/mythfrontend/playbackbox.cpp    |   37 +++++++++++++++++++++--
 mythtv/programs/mythfrontend/playbackbox.h      |    1 +
 mythtv/themes/MythCenter-wide/recordings-ui.xml |    8 +++--
 mythtv/themes/MythCenter/recordings-ui.xml      |    8 +++--
 7 files changed, 74 insertions(+), 6 deletions(-)

diff --git a/mythtv/libs/libmyth/programinfo.cpp b/mythtv/libs/libmyth/programinfo.cpp
index 89670c4..3bec015 100644
--- a/mythtv/libs/libmyth/programinfo.cpp
+++ b/mythtv/libs/libmyth/programinfo.cpp
@@ -4735,6 +4735,27 @@ void ProgramInfo::SaveInetRef(const QString &inet)
     SendUpdateEvent();
 }
 
+void ProgramInfo::SaveStars(uint u, uint range_max)
+{
+    stars = float(u) / float(range_max);
+
+    MSqlQuery query(MSqlQuery::InitCon());
+
+    query.prepare(
+        "UPDATE recorded "
+        "SET stars = :STARS "
+        "WHERE chanid = :CHANID AND starttime = :STARTTIME "
+        "AND recordid = :RECORDID");
+
+    query.bindValue(":STARS",      stars);
+    query.bindValue(":CHANID",     chanid);
+    query.bindValue(":STARTTIME",  recstartts);
+    query.bindValue(":RECORDID",   recordid);
+    query.exec();
+
+    SendUpdateEvent();
+}
+
 /** \brief Attempts to ascertain if the main file for this ProgramInfo
  *         is readable.
  *  \note This method often initiates a QUERY_CHECKFILE MythProto
diff --git a/mythtv/libs/libmyth/programinfo.h b/mythtv/libs/libmyth/programinfo.h
index ed06dc7..1aa1536 100644
--- a/mythtv/libs/libmyth/programinfo.h
+++ b/mythtv/libs/libmyth/programinfo.h
@@ -604,6 +604,7 @@ class MPUBLIC ProgramInfo
     void UpdateInUseMark(bool force = false);
     void SaveSeasonEpisode(uint seas, uint ep);
     void SaveInetRef(const QString &inet);
+    void SaveStars(uint u, uint range_max);
 
     // Extremely slow functions that cannot be called from the UI thread.
     QString DiscoverRecordingDirectory(void) const;
diff --git a/mythtv/libs/libmythtv/tv_play.cpp b/mythtv/libs/libmythtv/tv_play.cpp
index ac26d85..8006e1a 100644
--- a/mythtv/libs/libmythtv/tv_play.cpp
+++ b/mythtv/libs/libmythtv/tv_play.cpp
@@ -590,6 +590,10 @@ void TV::InitKeys(void)
             "Add Bookmark"), bkmKeys);
     REG_KEY("TV Playback", ACTION_TOGGLEBOOKMARK, QT_TRANSLATE_NOOP("MythControls",
             "Toggle Bookmark"), togBkmKeys);
+    REG_KEY("TV Frontend", "THMBUP",     QT_TRANSLATE_NOOP("MythControls",
+            "Increase rating"), "9");
+    REG_KEY("TV Frontend", "THMBDOWN",   QT_TRANSLATE_NOOP("MythControls",
+            "Decrease rating"), "7");
     REG_KEY("TV Playback", "BACK", QT_TRANSLATE_NOOP("MythControls",
             "Exit or return to DVD menu"), "Esc");
     REG_KEY("TV Playback", ACTION_MENUCOMPACT, QT_TRANSLATE_NOOP("MythControls",
diff --git a/mythtv/programs/mythfrontend/playbackbox.cpp b/mythtv/programs/mythfrontend/playbackbox.cpp
index 46c3c03..c5bdf7e 100644
--- a/mythtv/programs/mythfrontend/playbackbox.cpp
+++ b/mythtv/programs/mythfrontend/playbackbox.cpp
@@ -53,6 +53,7 @@
 #define LOC_ERR  QString("PlaybackBox Error: ")
 
 static const QString _Location = "Playback Box";
+const uint kMaxStars = 10;
 
 static int comp_programid(const ProgramInfo *a, const ProgramInfo *b)
 {
@@ -851,7 +852,7 @@ void PlaybackBox::UpdateUIListItem(MythUIButtonListItem *item,
 
     SetItemIcons(item, pginfo);
 
-    QString rating = QString::number(pginfo->GetStars(10));
+    QString rating = QString::number(pginfo->GetStars(kMaxStars));
 
     item->DisplayState(rating, "ratingstate");
 
@@ -959,7 +960,7 @@ void PlaybackBox::ItemLoaded(MythUIButtonListItem *item)
 
         item->DisplayState(state, "status");
 
-        item->DisplayState(QString::number(pginfo->GetStars(10)),
+        item->DisplayState(QString::number(pginfo->GetStars(kMaxStars)),
                            "ratingstate");
 
         SetItemIcons(item, pginfo);
@@ -3983,6 +3984,10 @@ bool PlaybackBox::keyPressEvent(QKeyEvent *event)
             showViewChanger();
         else if (action == "EDIT")
             EditScheduled();
+        else if (action == "THMBUP")
+            changeRating(true);
+        else if (action == "THMBDOWN")
+            changeRating(false);
         else if (m_titleList.size() > 1)
         {
             if (action == "DELETE")
@@ -5165,6 +5170,34 @@ void PlaybackBox::SetRecGroupPassword(const QString &newPassword)
     m_recGroupPwCache[m_recGroup] = newPassword;
 }
 
+void PlaybackBox::changeRating(bool increase)
+{
+    MythUIStateType *ratingState = dynamic_cast<MythUIStateType* >(
+                                        GetChild("ratingstate"));
+    if (!ratingState)
+        return;
+
+    if (GetFocusWidget() != m_recordingList)
+        return;
+
+    MythUIButtonListItem *item = m_recordingList->GetItemCurrent();
+    if (!item)
+        return;
+
+    ProgramInfo *pginfo = qVariantValue<ProgramInfo *>(item->GetData());
+    if (!pginfo)
+        return;
+
+    uint stars = pginfo->GetStars(kMaxStars);
+
+    if (increase && stars < kMaxStars)
+        pginfo->SaveStars(++stars, kMaxStars);
+    else if (!increase && stars > 0)
+        pginfo->SaveStars(--stars, kMaxStars);
+
+    ratingState->DisplayState(QString::number(stars));
+}
+
 ///////////////////////////////////////////////////
 
 GroupSelector::GroupSelector(MythScreenStack *lparent, const QString &label,
diff --git a/mythtv/programs/mythfrontend/playbackbox.h b/mythtv/programs/mythfrontend/playbackbox.h
index 74f5a56..e85c0cf 100644
--- a/mythtv/programs/mythfrontend/playbackbox.h
+++ b/mythtv/programs/mythfrontend/playbackbox.h
@@ -344,6 +344,7 @@ class PlaybackBox : public ScheduleCommon
     QString extract_job_state(const ProgramInfo &pginfo);
     QString extract_commflag_state(const ProgramInfo &pginfo);
 
+    void changeRating(bool increase);
 
     QRegExp m_prefixes;   ///< prefixes to be ignored when sorting
     QRegExp m_titleChaff; ///< stuff to remove for search rules
diff --git a/mythtv/themes/MythCenter-wide/recordings-ui.xml b/mythtv/themes/MythCenter-wide/recordings-ui.xml
index dd8613f..88e58a3 100644
--- a/mythtv/themes/MythCenter-wide/recordings-ui.xml
+++ b/mythtv/themes/MythCenter-wide/recordings-ui.xml
@@ -277,6 +277,10 @@
             </statetype>
         </buttonlist>
 
+        <statetype name="ratingstate" from="baseratingstate">
+            <position>30,660</position>
+        </statetype>
+
         <imagetype name="autoexpire">
             <filename>autoexpire.png</filename>
             <position>988,660</position>
@@ -400,11 +404,11 @@
         </textarea>
 
         <textarea name="description" from="basetextarea">
-            <area>30,480,885,210</area>
+            <area>30,480,885,180</area>
             <multiline>yes</multiline>
             <scroll direction="vertical" />
             <cutdown>no</cutdown>
-            <template>%"|SUBTITLE|" %%(|STARS|) %%DESCRIPTION%</template>
+            <template>%"|SUBTITLE|" %%DESCRIPTION%</template>
         </textarea>
 
         <imagetype name="preview" depends="!coverart">
diff --git a/mythtv/themes/MythCenter/recordings-ui.xml b/mythtv/themes/MythCenter/recordings-ui.xml
index 6aaaf22..f95fd4c 100644
--- a/mythtv/themes/MythCenter/recordings-ui.xml
+++ b/mythtv/themes/MythCenter/recordings-ui.xml
@@ -277,6 +277,10 @@
             </statetype>
         </buttonlist>
 
+        <statetype name="ratingstate" from="baseratingstate">
+            <position>20,560</position>
+        </statetype>
+
         <imagetype name="autoexpire">
             <filename>autoexpire.png</filename>
             <position>630,518</position>
@@ -401,11 +405,11 @@
         </textarea>
 
         <textarea name="description" from="basetextarea">
-            <area>18,425,572,160</area>
+            <area>18,425,572,135</area>
             <multiline>yes</multiline>
             <scroll direction="vertical" />
             <cutdown>no</cutdown>
-            <template>%"|SUBTITLE|" %%(|STARS|) %%DESCRIPTION%</template>
+            <template>%"|SUBTITLE|" %%DESCRIPTION%</template>
         </textarea>
 
         <imagetype name="preview" depends="!coverart">
-- 
1.7.9.5

